# Next cell is for running on GPU
開啟檔案後預設的運行單元為CPU，若要改成GPU則可以按下面兩圖操作：https://mattwang44.github.io/en/images/DL/colab/5.png,
https://mattwang44.github.io/en/images/DL/colab/6.png

更改後可以用運行這一段程式碼確認GPU運行正常

In [1]:
import tensorflow as tf
device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
  raise SystemError('GPU device not found')
print('Found GPU at: {}'.format(device_name))

Found GPU at: /device:GPU:0


# Next cell is for using TPU.
See: https://colab.research.google.com/notebooks/tpu.ipynb#scrollTo=G2pjP0c5VqhK

In [0]:
import os
import pprint
%tensorflow_version 1.x
import tensorflow as tf

if 'COLAB_TPU_ADDR' not in os.environ:
  print('ERROR: Not connected to a TPU runtime; please see the first cell in this notebook for instructions!')
else:
  tpu_address = 'grpc://' + os.environ['COLAB_TPU_ADDR']
  print ('TPU address is', tpu_address)

  with tf.Session(tpu_address) as session:
    devices = session.list_devices()
    
  print('TPU devices:')
  pprint.pprint(devices)

運行下面的程式碼，會出現鏈接，點開鏈接，選擇自己的谷歌帳號，將得到的代碼輸入鏈接下面的框中即可。

In [2]:
!apt-get install -y -qq software-properties-common python-software-properties module-init-tools
!add-apt-repository -y ppa:alessandro-strada/ppa 2>&1 > /dev/null
!apt-get update -qq 2>&1 > /dev/null
!apt-get -y install -qq google-drive-ocamlfuse fuse
from google.colab import auth
auth.authenticate_user()
from oauth2client.client import GoogleCredentials
creds = GoogleCredentials.get_application_default()
import getpass
!google-drive-ocamlfuse -headless -id={creds.client_id} -secret={creds.client_secret} < /dev/null 2>&1 | grep URL
vcode = getpass.getpass()
!echo {vcode} | google-drive-ocamlfuse -headless -id={creds.client_id} -secret={creds.client_secret}

E: Package 'python-software-properties' has no installation candidate
Selecting previously unselected package google-drive-ocamlfuse.
(Reading database ... 145605 files and directories currently installed.)
Preparing to unpack .../google-drive-ocamlfuse_0.7.14-0ubuntu1~ubuntu18.04.1_amd64.deb ...
Unpacking google-drive-ocamlfuse (0.7.14-0ubuntu1~ubuntu18.04.1) ...
Setting up google-drive-ocamlfuse (0.7.14-0ubuntu1~ubuntu18.04.1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Please, open the following URL in a web browser: https://accounts.google.com/o/oauth2/auth?client_id=32555940559.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&response_type=code&access_type=offline&approval_prompt=force
··········
Please, open the following URL in a web browser: https://accounts.google.com/o/oauth2/auth?client_id=32555940559.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope

In [3]:
!nvidia-smi # this will work if run on GPU, but fail if run on TPU

Mon Nov 25 01:56:09 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
| N/A   73C    P0    77W / 149W |     69MiB / 11441MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

掛載谷歌雲端硬碟：以下程式碼則可以幫我們把虛擬機和自己的Google Drive連接，執行後虛擬機目錄下會多出一個名為Drive(可自行更改)*的資料夾，內容即為自己的雲端資料。*

In [0]:
!mkdir -p Drive 
!google-drive-ocamlfuse Drive -o nonempty

每次修改了雲盤中的文件以後都最好運行一次，運行了沒有任何提示便成功了。

若是出現Warning: apt-key output should not be parsed (stdout is not a terminal)警告標語，代表和自己雲端連接的權限已經到期，重新執行上面這段code就可以重新連接。

設置工作路徑：

In [0]:
import os 
os.chdir( 'Drive' )

In [6]:
!ls 指導的學生/2019_吳玟萱/2019_1104_Joke_Datasets

 cache_dir
 Excel2txt.py
 interpreting-text-classification-models-with-eli5.ipynb
 JokeHumorLevel.ipynb
 JokeMotivation.ipynb
 JokeSkill.ipynb
 mlabel_corpora
 Multi-Label-Text-Classification-master
 out
 __pycache__
 runs
 SimpleTransformer_HumorLevel.ipynb
 SimpleTransformer_Motivation_colab.ipynb
 SimpleTransformer_Skill_colab.ipynb
 simpletransformers-master
 SNsanafonGyou.ttf
 Stopwords.py
'~$笑話標記.xlsx'
 笑話標記.xlsx
 笑話資料庫.xlsx
 負範例.xlsx


In [0]:
!cd 指導的學生/2019_吳玟萱/2019_1104_Joke_Datasets
!pwd

/content/Drive


In [0]:
#!pip install pytorch cudatoolkit=10.0 -c pytorch

In [0]:
!pip show pytorch

In [0]:
!pip --version

In [7]:
!pip install transformers

     |████████████████████████████████| 317kB 6.7MB/s 
     |████████████████████████████████| 1.0MB 42.8MB/s 
     |████████████████████████████████| 645kB 49.7MB/s 
     |████████████████████████████████| 860kB 42.0MB/s 
  Created wheel for sacremoses: filename=sacremoses-0.0.35-cp36-none-any.whl size=883999 sha256=8fd08bb92aa3f103e20bf7d31431bb72f8f84371adb2b32f653ed28644ed1227
  Stored in directory: /root/.cache/pip/wheels/63/2a/db/63e2909042c634ef551d0d9ac825b2b0b32dede4a6d87ddc94
Successfully built sacremoses


In [8]:
!pip install seqeval

  Created wheel for seqeval: filename=seqeval-0.0.12-cp36-none-any.whl size=7424 sha256=610d41e9906a9de4732510a7e5cc2a95c78dae78c2d00f2d92f462997838178d
  Stored in directory: /root/.cache/pip/wheels/4f/32/0a/df3b340a82583566975377d65e724895b3fad101a3fb729f68
Successfully built seqeval


In [9]:
!pip install tensorboardx

     |████████████████████████████████| 194kB 6.4MB/s 


In [10]:
!pip install simpletransformers

     |████████████████████████████████| 71kB 4.3MB/s 


In [0]:
from simpletransformers.classification import MultiLabelClassificationModel
import pandas as pd
import sys, time

In [0]:
data_path = "指導的學生/2019_吳玟萱/2019_1104_Joke_Datasets/mlabel_corpora/JokeMotivation.txt"

In [13]:
# set global variables: df
df = pd.read_csv(data_path, delimiter="\t")
#data_raw = df.loc[np.random.choice(data_raw.index, size=2000)]
print(df.shape) # same as data_raw.shape in Jupyter

(3365, 9)


In [14]:
from sklearn.model_selection import train_test_split

# ID=L1850 為分界，之前：吳玟萱，之後：黃亭筠，均為中文系同一屆
train, test = train_test_split(df, train_size=1691, shuffle=False) 
# (tempararily) set global variables: train, test 
print(train.shape)
print(test.shape)
print(train.columns)

(1691, 9)
(1674, 9)
Index(['ID', 'Title', 'Content', 'Affinity', 'Self_Improvement', 'Attack',
       'Self_Depression', 'Taboo', 'Others'],
      dtype='object')


In [15]:
# set global variables: train_df, test_df
# https://stackoverflow.com/questions/34682828/extracting-specific-selected-columns-to-new-dataframe-as-a-copy
# selected column index: 1, 6, 7
# new = old.iloc[: , [1, 6, 7]].copy() 
# new = old[['A', 'C', 'D']].copy()

train_df = train[['Content']].copy()
#train_df = train_df.drop(labels = ['ID', 'Title', 'Content'], axis=1)
#https://stackoverflow.com/questions/33098383/merge-multiple-column-values-into-one-column-in-python-pandas
train_df['labels'] = train[train.columns[3:]].apply(
    lambda x: list(x), axis=1)
#    lambda x: ','.join(x.dropna().astype(str)), axis=1)
#train_df['labels'] = train_df[['labels']].apply(lambda x: x.split(','), axis=1)
# train.columns = ['ID', 'Title', 'Content', 'Pun', 'Exaggeration', 'Anthropomorphism', 'Bridge_Inference', 'Illogical', 'Irony', 'Imitation', 'Others']
#train_df = train_df.drop(labels = train.columns, axis=1)

test_df = test.loc[:, ['Content']].copy()
#https://stackoverflow.com/questions/33098383/merge-multiple-column-values-into-one-column-in-python-pandas
test_df['labels'] = test.loc[:, test.columns[3:]].apply(
    lambda x: list(x), axis=1)
#    lambda x: ','.join(x.dropna().astype(str)), axis=1)
#test_df['labels'] = test_df[['labels']].apply(lambda x: x.split(','), axis=1)
#test_df = test_df.drop(labels = test.columns, axis=1)

print(f'{train.shape}, {test.shape}, {train_df.shape}, {test_df.shape}')
print()
print(train_df.head())
print(test_df.head())

(1691, 9), (1674, 9), (1691, 2), (1674, 2)

                                             Content              labels
0  員工：老闆，您必須幫我加薪，已經有三家公司在找我了！     老闆：哪三家？     員工：...  [0, 0, 0, 1, 0, 0]
1  某市政府辦公大樓落成，門口缺副對聯。     副市長揮毫     上聯：說實話辦實事一身正氣...  [0, 0, 1, 0, 0, 0]
2  中午老闆視察自己的建築工地時，發現有個人在角落玩手機。     老闆：你月薪多少？     ...  [0, 0, 0, 1, 0, 0]
3  一天，一位法官的女友看見兩個蚊子，便叫法官打死。     只見法官只把那個肚子飽飽的蚊子打死...  [0, 0, 0, 0, 0, 1]
4  辦公室中兩位女同事吵起來了。     經理忍無可忍：「太不像話了！現在是什麼情況？你們把原因...  [0, 0, 1, 0, 0, 0]
                       Content              labels
1691  即使你已名花有主，我也要移花接木把你移到我身邊。  [0, 0, 0, 0, 0, 1]
1692              錢不是問題，問題是沒錢！  [0, 0, 0, 1, 0, 0]
1693    自從我變成了狗屎，就再也沒有人踩在我頭上了。  [0, 0, 0, 1, 0, 0]
1694         蘋果最光輝的一刻就是砸在牛頓頭上！  [0, 0, 0, 0, 0, 1]
1695   其實上課睡覺不是我們的錯，只怪地球吸引力太大。  [0, 0, 0, 0, 0, 1]


In [16]:
# Create a MultiLabelClassificationModel
model = MultiLabelClassificationModel('bert', 'bert-base-chinese', 
#model = MultiLabelClassificationModel('bert', '/Users/sam/GoogleDrive/BERT/bert/chinese_L-12_H-768_A-12/', 
    num_labels=6, use_cuda=True, 
    args={'reprocess_input_data':True, 'overwrite_output_dir':True, 'num_train_epochs':10,
          'max_seq_length': 512, 'train_batch_size': 8, 'eval_batch_size': 8, 
          'output_dir':'outputs/Motivation'})
# You can set class weights by using the optional weight argument


100%|██████████| 411577189/411577189 [00:15<00:00, 27098360.27B/s]


In [17]:
# https://ipywidgets.readthedocs.io/en/stable/user_install.html
# conda update -n base -c defaults conda # this will update conda version for all virtualenv
# conda install -c conda-forge ipywidgets

stime = time.time()
# Train the model
model.train_model(train_df)
print('training time: %1.2f'%(time.time()-stime))


Converting to features started.


Running loss: 0.249413

Running loss: 0.233116

Running loss: 0.354023

Running loss: 0.017526

Running loss: 0.078948

Running loss: 0.106738

Running loss: 0.010444

Running loss: 0.010128

Running loss: 0.211466

Running loss: 0.005026
Training of bert model complete. Saved to outputs/Motivation.
training time: 3627.96


In [18]:
stime = time.time()
# Evaluate the model
result, model_outputs, wrong_predictions = model.eval_model(test_df)
print(result)
print(model_outputs) # this is the raw prediction
print('Wrong Prediction:')
print(wrong_predictions)
print('Eval time: %1.2f'%(time.time()-stime))

Converting to features started.



{'LRAP': 0.7070390282755902, 'eval_loss': 0.49969153045454906}
[[0.02280636 0.04897381 0.03251348 0.89226806 0.03164655 0.05865327]
 [0.01699421 0.03553129 0.01651633 0.8368093  0.02144256 0.13680331]
 [0.0038724  0.01180652 0.93066525 0.03568676 0.01831111 0.03437178]
 ...
 [0.00334068 0.00302795 0.00413219 0.0029388  0.00642982 0.99008876]
 [0.0037867  0.00272131 0.00481092 0.00256829 0.00537018 0.9894309 ]
 [0.00366253 0.01504941 0.01185209 0.049722   0.00852506 0.8570056 ]]
Wrong Prediction:
[<simpletransformers.classification.classification_utils.InputExample object at 0x7f8eef2b99e8>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f8e4674c400>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f8e4674c470>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f8e4674c4a8>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f8e4674c358>, <simpletransforme

In [0]:
from sklearn.metrics import accuracy_score, roc_auc_score, roc_curve
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
import numpy as np 

In [0]:
# define the evaluation metrics
def tcfunc(x, n=4): # trancate a number to have n decimal digits
    d = '0' * n
    d = int('1' + d)
# https://stackoverflow.com/questions/4541155/check-if-a-number-is-int-or-float
    if isinstance(x, (int, float)): return int(x * d) / d
    return x

def print_cls_report(y_true, prediction):
    print('Test accuracy is %1.4f'%(accuracy_score(y_true, prediction)))

    print(classification_report(y_true, prediction))
    
    # http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html
    print("\tPrecision\tRecall\tF1\tSupport")
    (Precision, Recall, F1, Support) = list(map(tcfunc, 
        precision_recall_fscore_support(y_true, prediction, average='micro')))
    print("Micro\t{}\t{}\t{}\t{}".format(Precision, Recall, F1, Support))

    (Precision, Recall, F1, Support) = list(map(tcfunc, 
        precision_recall_fscore_support(y_true, prediction, average='macro')))
    print("Macro\t{}\t{}\t{}\t{}".format(Precision, Recall, F1, Support))
    
#    if True:
    if False:
        print(confusion_matrix(y_true, prediction))
        try: 
            print(classification_report(y_true, prediction, digits=4))
        except ValueError:
            print('May be some category has no predicted samples')
        show_confusion_matrix(prediction)


    print(f'y_true.shape={y_true.shape}, prediction.shape={prediction.shape}')
    #print(y_true.head())
    #print(prediction[0:6])

    # https://stackoverflow.com/questions/152580/whats-the-canonical-way-to-check-for-type-in-python
    pred = prediction
    if not isinstance(pred, np.ndarray): pred = prediction.toarray()
    print(type(y_true), type(prediction), type(pred))
    try:
        # https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html
        print('macro roc_auc_score is %1.4f'%(roc_auc_score(y_true, pred, average='macro'))) # default average=’macro’
        print('micro roc_auc_score is %1.4f'%(roc_auc_score(y_true, pred, average='micro')))
    except:
        print("roc_auc_score error!!!")
    try:
        fpr, tpr, thresholds = roc_curve(y_true, pred)
        print(f'fpr={fpr}\ntpr={tpr}\nthresholds={thresholds}')
    except:
        print('roc_curve error!!!')

In [21]:
type(model_outputs)
# https://stackoverflow.com/questions/35215161/most-efficient-way-to-map-function-over-numpy-array

# Obtain array of square of each element in x
threshold = lambda x: 1 if x>0.5 else 0 
predictions = np.array([[threshold(x) for x in row] for row in model_outputs])
print(type(predictions), type(test_df['labels']))
print('predictions:')
print(predictions)
y_true = np.array([[x for x in row] for row in test_df['labels'] ])
print(f'y_true, {type(y_true)}')
print(y_true)

<class 'numpy.ndarray'> <class 'pandas.core.series.Series'>
predictions:
[[0 0 0 1 0 0]
 [0 0 0 1 0 0]
 [0 0 1 0 0 0]
 ...
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]]
y_true, <class 'numpy.ndarray'>
[[0 0 0 0 0 1]
 [0 0 0 1 0 0]
 [0 0 0 1 0 0]
 ...
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]]


In [22]:
print_cls_report(y_true, predictions)

Test accuracy is 0.5119
              precision    recall  f1-score   support

           0       0.43      0.16      0.23        19
           1       0.00      0.00      0.00        16
           2       0.21      0.50      0.30       143
           3       0.07      0.29      0.12        17
           4       0.37      0.48      0.42       229
           5       0.84      0.55      0.67      1254

   micro avg       0.57      0.53      0.55      1678
   macro avg       0.32      0.33      0.29      1678
weighted avg       0.70      0.53      0.58      1678
 samples avg       0.52      0.52      0.52      1678

	Precision	Recall	F1	Support
Micro	0.5746	0.525	0.5487	None
Macro	0.3203	0.3301	0.288	None
y_true.shape=(1674, 6), prediction.shape=(1674, 6)
<class 'numpy.ndarray'> <class 'numpy.ndarray'> <class 'numpy.ndarray'>
macro roc_auc_score is 0.6102
micro roc_auc_score is 0.7235
roc_curve error!!!


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in samples with no predicted labels.
  'precision', 'predicted', average, warn_for)
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/classification.py:1439: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in samples with no true labels.
  'recall', 'true', average, warn_for)


# When epochs=10 wit P100 GPU
  1. Running loss: 0.378187
  2. Running loss: 0.159992
  3. Running loss: 0.129692
  4. Running loss: 0.019266
  5. Running loss: 0.008311
  6. Running loss: 0.206701
  7. Running loss: 0.012348
  8. Running loss: 0.005116
  9. Running loss: 0.064961
  10. Running loss: 0.004230

  Current iteration
  100% 212/212 [02:09<00:00, 2.42it/s]

  Training of bert model complete. 
  Saved to outputs/Motivation.

training time: 1138.04

Test accuracy is 0.4857

              precision    recall  f1-score   support
           0       0.38      0.16      0.22        19
           1       0.00      0.00      0.00        16
           2       0.19      0.52      0.27       143
           3       0.22      0.24      0.23        17
           4       0.33      0.51      0.40       229
           5       0.85      0.51      0.64      1254

    micro avg       0.54      0.50      0.52      1678
    macro avg       0.33      0.32      0.29      1678
    weighted avg       0.70      0.50      0.56      1678
    samples avg       0.49      0.50      0.50      1678

	Precision	Recall	F1	Support
    Micro	0.5448	0.4994	0.5211	None
    Macro	0.3262	0.3219	0.2934	None

y_true.shape=(1674, 6), prediction.shape=(1674, 6)

    macro roc_auc_score is 0.6055
    micro roc_auc_score is 0.7079
roc_curve error!!!


# When epochs=10 wit K80 GPU
    1. Running loss: 0.249413
    2. Running loss: 0.233116
    3. Running loss: 0.354023
    4. Running loss: 0.017526
    5. Running loss: 0.078948
    6. Running loss: 0.106738
    7. Running loss: 0.010444
    8. Running loss: 0.010128
    9. Running loss: 0.211466
    10. Running loss: 0.005026
100% 212/212 [06:26<00:00, 1.38s/it]

Training of bert model complete. 
Saved to outputs/Motivation.

training time: 3627.96



In [0]:
stime = time.time()
predictions, raw_outputs = model.predict(['即使你已名花有主，我也要移花接木把你移到我身邊。', '錢不是問題，問題是沒錢！'])
print(predictions)
print(raw_outputs)
print('Predict time: %1.2f'%(time.time()-stime))

Converting to features started.



[[0, 0, 0, 1, 0, 0], [0, 0, 0, 1, 0, 0]]
[[0.17519423 0.02841783 0.09872122 0.5164647  0.00792587 0.01830811]
 [0.03143917 0.04437851 0.04562189 0.9163727  0.02270437 0.08641115]]
Predict time: 1.18


# When epochs=4
batch_size=8, length=512

    1. Running loss: 0.406653
    2. Running loss: 0.147888
    3.Running loss: 0.125427
    4. Running loss: 0.096597
    100% 212/212 [01:45<00:00, 2.39it/s]
    Training of bert model complete. 
    Saved to outputs/Motivation.

training time: 466.28

Test accuracy is 0.4851

              precision    recall  f1-score   support
           0       0.50      0.11      0.17        19
           1       0.00      0.00      0.00        16
           2       0.22      0.50      0.30       143
           3       0.00      0.00      0.00        17
           4       0.33      0.60      0.42       229
           5       0.86      0.49      0.63      1254

    micro avg       0.56      0.49      0.53      1678
    macro avg       0.32      0.28      0.25      1678
    weighted avg       0.71      0.49      0.55      1678
    samples avg       0.49      0.49      0.49      1678

	Precision	Recall	F1	Support
    Micro	0.5636	0.4934	0.5262	None
    Macro	0.3176	0.2827	0.2543	None
    y_true.shape=(1674, 6), prediction.shape=(1674, 6)
    macro roc_auc_score is 0.5913
    micro roc_auc_score is 0.7084
    roc_curve error!!!